File: backend/chat_handler.py
"""Always route user questions through Glean MCP."""

from __future__ import annotations

import re
from typing import Any, AsyncIterator

from mcp_client import GleanMcpClient

_FOOTNOTE_DEF = re.compile(r"^\[\^(\d+)\]:\s*\[([^\]]*)\]\(([^)]+)\)\s*$", re.MULTILINE)
_CHAT_METADATA = re.compile(r"\n---\nchatId:.*", re.DOTALL)
_FOLLOWUP_BLOCK = re.compile(
    r"\n---\n(?!chatId:)(.+?)(?=\n\[\^\d+\]:|\n---\nchatId:|$)",
    re.DOTALL,
)


def _parse_response(text: str) -> tuple[str, list[dict[str, str]]]:
    citations: list[dict[str, str]] = []
    for match in _FOOTNOTE_DEF.finditer(text):
        citations.append(
            {"id": match.group(1), "title": match.group(2).strip(), "url": match.group(3).strip()}
        )

    body = _CHAT_METADATA.sub("", text)
    body = _FOOTNOTE_DEF.sub("", body)
    body = _FOLLOWUP_BLOCK.sub("", body)
    body = re.sub(r"\n{3,}", "\n\n", body).strip()
    return body, citations


def _build_chat_args(tool_name: str, user_message: str, history: list[dict[str, str]]) -> dict[str, Any]:
    if tool_name == "chat":
        context = [
            f"{m['role']}: {m['content']}"
            for m in history
            if m["role"] in ("user", "assistant") and m["content"].strip()
        ]
        args: dict[str, Any] = {"message": user_message}
        if context:
            args["context"] = context
        return args

    if tool_name in ("search", "glean_search"):
        return {"query": user_message}

    return {"message": user_message, "query": user_message}


async def handle_chat(
    mcp: GleanMcpClient,
    messages: list[dict[str, str]],
) -> AsyncIterator[dict[str, Any]]:
    user_messages = [m for m in messages if m["role"] == "user"]
    if not user_messages:
        yield {"type": "error", "content": "No user message provided."}
        return

    user_message = user_messages[-1]["content"].strip()
    if not user_message:
        yield {"type": "error", "content": "Empty message."}
        return

    prior = messages[:-1]
    tool_name = mcp.resolve_tool()

    args = _build_chat_args(tool_name, user_message, prior)
    yield {"type": "status", "content": "Asking Glean..."}
    yield {"type": "tool_call", "name": tool_name, "arguments": args}

    try:
        tool_result = await mcp.call_tool(tool_name, args)
    except Exception as exc:
        yield {"type": "error", "content": f"MCP tool failed: {exc}"}
        return

    yield {
        "type": "tool_result",
        "name": tool_name,
        "content": tool_result.text,
        "raw": tool_result.raw,
    }
    answer, citations = _parse_response(tool_result.text)
    yield {"type": "text", "content": answer, "citations": citations}

File: backend/src/chat-handler.ts
import type { GleanMcpClient } from "./mcp-client.js";

const FOOTNOTE_DEF = /^\[\^(\d+)\]:\s*\[([^\]]*)\]\(([^)]+)\)\s*$/gm;
const CHAT_METADATA = /\n---\nchatId:.*/s;
const FOLLOWUP_BLOCK = /\n---\n(?!chatId:)(.+?)(?=\n\[\^\d+\]:|\n---\nchatId:|$)/s;

export interface ChatMessage {
  role: string;
  content: string;
}

export interface ChatEvent {
  type: string;
  content?: string;
  name?: string;
  arguments?: Record<string, unknown>;
  raw?: Record<string, unknown>;
  citations?: Array<{ id: string; title: string; url: string }>;
}

function parseResponse(text: string): [
  string,
  Array<{ id: string; title: string; url: string }>,
] {
  const citations: Array<{ id: string; title: string; url: string }> = [];
  for (const match of text.matchAll(FOOTNOTE_DEF)) {
    citations.push({
      id: match[1],
      title: match[2].trim(),
      url: match[3].trim(),
    });
  }

  let body = text.replace(CHAT_METADATA, "");
  body = body.replace(FOOTNOTE_DEF, "");
  body = body.replace(FOLLOWUP_BLOCK, "");
  body = body.replace(/\n{3,}/g, "\n\n").trim();
  return [body, citations];
}

function buildChatArgs(
  toolName: string,
  userMessage: string,
  history: ChatMessage[],
): Record<string, unknown> {
  if (toolName === "chat") {
    const context = history
      .filter((m) => (m.role === "user" || m.role === "assistant") && m.content.trim())
      .map((m) => `${m.role}: ${m.content}`);
    const args: Record<string, unknown> = { message: userMessage };
    if (context.length > 0) args.context = context;
    return args;
  }

  if (toolName === "search" || toolName === "glean_search") {
    return { query: userMessage };
  }

  return { message: userMessage, query: userMessage };
}

export async function* handleChat(
  mcp: GleanMcpClient,
  messages: ChatMessage[],
): AsyncGenerator<ChatEvent> {
  const userMessages = messages.filter((m) => m.role === "user");
  if (userMessages.length === 0) {
    yield { type: "error", content: "No user message provided." };
    return;
  }

  const userMessage = userMessages[userMessages.length - 1].content.trim();
  if (!userMessage) {
    yield { type: "error", content: "Empty message." };
    return;
  }

  const prior = messages.slice(0, -1);
  const toolName = mcp.resolveTool();
  const args = buildChatArgs(toolName, userMessage, prior);

  yield { type: "status", content: "Asking Glean..." };
  yield { type: "tool_call", name: toolName, arguments: args };

  let toolResult;
  try {
    toolResult = await mcp.callTool(toolName, args);
  } catch (err) {
    yield { type: "error", content: `MCP tool failed: ${err}` };
    return;
  }

  yield {
    type: "tool_result",
    name: toolName,
    content: toolResult.text,
    raw: toolResult.raw,
  };

  const [answer, citations] = parseResponse(toolResult.text);
  yield { type: "text", content: answer, citations };
}
